fixes and optimizations

hankedan000 opened the pull request:
- shared the CutPieces scene between the BeepCube and ChainLink scenes
- CutPieces are now pre-instanced and live inside of their parent cube's scene tree (as opposed to reparented into the the track)
- Cube slice particles are no longer dynamically instance (pre-instanced when cube is made and live in cube's scene tree, similar to the way CutPieces are managed)
- rewrote a lot of the ScenePool logic. it was becoming way to specific to cube types. could potentially pool Bombs and Walls too, but performance seems ok so far ...
- cleaned up old materials that are no longer used
- fixed issue with the UIRayCast becoming offset independent of saber custom offsets/rotations
- reduced time consumed by useless rumble API calls
- improved performance of the saber tail history management (also made that code more readable by reducing triple-nested list data structure)

hankedan000 committed:
use "Local Floor" reference space

was previous set to "Local" which doesn't support recentering (keeps player head fixed in virtual world)

hankedan000 committed:
improve cube cutting performance

 * pre-instance the CutPieces objects
 * fixed bugs in the ScenePool leaking cubes on release()

hankedan000 committed:
seems like this is a stale/unused setting?

i have the renderer drop down to "Mobile". you only acquire this additional setting when you switch to "Compatibility" rendere, but then this setting gets removed when you switch back to "Mobile", so i think it should be removed.

hankedan000 committed:
Fixes race condition in BackgroundImgLoader

hankedan000 committed:
Fix unecessary repeated calls to trigger_haptic_pulse()

i'm not sure if the old API needed a call to trigger_haptic_pulse() with intesity=0.0 in order to terminate the rumble, but the current API has a duration argument. repeatedly calling trigger_haptic_pulse() seems to have no gain besides a ~0.25ms cost per call.

hankedan000 committed:
Fixes bug in impulse vector direction

if the cut plane was vertical, then the cut pieces were flying up/down instead of left/right

hankedan000 committed:
Revert "pool cube cut materials to reuse and improve performance"

This reverts commit a2c5b81a25aec884db582d1abfd0ed5aef1710a0.

hankedan000 committed:
Fixes GlobalReference autoload

hankedan000 committed:
ChainLinks are pooled now. It also shares the CutPiece class with BeepCube

hankedan000 committed:
2nd pass at pooling chains and cubes

achieving a very stable 90FPS on Quest 2 now. maps that use bombs and walls (not pooled yet) still suffer from some frame drops when those scenes are instanced and added to track

hankedan000 committed:
removed unused cube shader and materials

hankedan000 committed:
SliceParticles are no longer dynamically spawned

hankedan000 committed:
Created BeepCubeTest scene

hankedan000 committed:
Improved performance of tail.gd:_process()

introduces a HistoricalPosition class to improve readability and reduce the usage of arrays of arrays of arrays (fewer allocs)

hankedan000 committed:
Fixed UIRaycast being misaligned with lightsaber manual offsets

shocktail39 commented:
oh shoot, an early contributor.  welcome back, and nice to meet you!  i would like to ask about one thing: in commit [improve cube cutting performance](https://github.com/leandrodreamer/BeepSaber/pull/22/commits/99e7897751f6a0071dbd3beddba1fb1d544b7ef0), you replaced a pop_back with a pop_front in the scene pool class's acquire function.  is there a particular reason for this?  according to [the godot documentation](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-method-pop-front), pop_front takes longer than pop_back due to it having to move every remaining element one index over, with how long it takes depending on the size of the array.

hankedan000 commented:
> oh shoot, an early contributor. welcome back, and nice to meet you! i would like to ask about one thing: in commit [improve cube cutting performance](https://github.com/leandrodreamer/BeepSaber/pull/22/commits/99e7897751f6a0071dbd3beddba1fb1d544b7ef0), you replaced a pop_back with a pop_front in the scene pool class's acquire function. is there a particular reason for this? according to [the godot documentation](https://docs.godotengine.org/en/stable/classes/class_array.html#class-array-method-pop-front), pop_front takes longer than pop_back due to it having to move every remaining element one index over, with how long it takes depending on the size of the array.

Glad to be back, and glad to see the project live on in Godot 4.

Good observation. I don't think that change is necessary anymore, and I will revert that line and retest.

That change was a bit of a "hack" to make sure that released scenes take a while to cycle back to getting acquired again. In that commit, I made the cut piece objects managed by their parent cube. Originally, the cube would release() itself immediate after being cut, instead of waiting for its cut pieces to die off too. If the cubes were immediate pushed back onto the resource pool, this would make them immediate get reused again before their cut pieces had finished animating themselves.

I have since changed the way the cube release() logic works [(08f8418)](https://github.com/leandrodreamer/BeepSaber/pull/22/commits/08f8418aca733ca8080a115c7831503cb431942f#diff-2e55e052d8925021642e9f494b6a7be812b643ef8051e3900d908b49d0493b7aR27-R28) and this pop_front() should no longer be necessary.

hankedan000 committed:
no need to pop from front anymore

this was a hack to make it so cubes didn't get reused until after their cut pieces has died off. the cube releasing logic has since been udpated to track when both cut pieces have died.

shocktail39 commented:
ah good catch on the pieces!

leandrodreamer commented:
Hi, sorry for not giving any sign of life around here, been very busy, I'll try to review this soon-ish
